library/include/Core.h

```diff
@@ -353,47 +354,60 @@ namespace DFHack
         friend struct CoreSuspendReleaseMain;
     };
 
-    class CoreSuspenderBase  : protected std::unique_lock<std::recursive_mutex> {
+    class CoreSuspenderBase : protected std::unique_lock< decltype(Core::CoreSuspendMutex) > {
     protected:
-        using parent_t = std::unique_lock<std::recursive_mutex>;
+        using mutex_type = decltype(Core::CoreSuspendMutex);
+        using parent_t = std::unique_lock< mutex_type >;
         std::thread::id tid;
+        Core& core;
 
-        CoreSuspenderBase(std::defer_lock_t d) : CoreSuspenderBase{&Core::getInstance(), d} {}
+        CoreSuspenderBase() : CoreSuspenderBase{ Core::getInstance() } {}
 
-        CoreSuspenderBase(Core* core, std::defer_lock_t) :
-            /* Lock the core */
-            parent_t{core->CoreSuspendMutex,std::defer_lock},
+        CoreSuspenderBase(Core& core) :
+            /* Lock the core (jk not really) */
+            parent_t{core.CoreSuspendMutex,std::defer_lock},
             /* Mark this thread to be the core owner */
-            tid{}
+            tid{},
+            core{ core }
         {}
+
     public:
         void lock()
         {
-            auto& core = Core::getInstance();
             parent_t::lock();
-            tid = core.ownerThread.exchange(std::this_thread::get_id(),
-                    std::memory_order_acquire);
+            complete_lock();
+        }
+
+        bool try_lock()
+        {
+            bool locked = parent_t::try_lock_for(std::chrono::milliseconds(100));
+            if (locked) complete_lock();
+            return locked;
         }
 
         void unlock()
         {
-            auto& core = Core::getInstance();
+            if (!owns_lock())
+                return;
             /* Restore core owner to previous value */
             core.ownerThread.store(tid, std::memory_order_release);
             if (tid == std::thread::id{})
                 Lua::Core::Reset(core.getConsole(), "suspend");
             parent_t::unlock();
         }
 
-        bool owns_lock() const noexcept
-        {
-            return parent_t::owns_lock();
+        ~CoreSuspenderBase() {
+            unlock();
         }
 
-        ~CoreSuspenderBase() {
-            if (owns_lock())
-                unlock();
+    protected:
+        void complete_lock()
+        {
+            auto& core = Core::getInstance();
```

Comment:
why do we need to get another core here (shadowing the class field)?

Reply:
there were a bunch of these i removed, i just missed that one in the refactor. thanx